new/resources-anywhere.md

```diff
@@ -0,0 +1,211 @@
+
+LWRPs created via `resources/x.rb` in a cookbook are also exempt from `TopLevelConstants`.
+
+#### Name Conflicts
+
+Resources can try to take the same name in recipe DSL (such as when two cookbooks both implement a similar thing).  Fact of life.  Presently when two custom resources try to take the same name, presently they will do one of two unsatisfactory things:
+
+- Raise a confusing "subclassing the wrong class" error on load; or
+- Silently overwrite and merge with one another, with the last one winning
+
+Now, when you are in separate namespaces, the first thing simply won't happen.
+
+We will avoid the second thing by checking whether the method `my_resource` is already defined, and if it is, we
+
+#### Replaced Classes
+
+It is possible to replace a class in Ruby (you can simply go assign ).  Due to the way this is implemented, the new class will in fact be used (which is sort of what a user would expect).  However, it's janky and something we'd like to warn about.  When the user attempts to use the resource (by typing `my_resource`), we will check in `build_resource` that the class has not changed from what was originally registered, and emit a warning that the class isn't what we expect and that they should contact the author of the new class to get the warning fixed.
+
+To *legitimately* do this, you should use `chef_recipe_dsl` instead.
+
+#### `resources` conflicts between cookbooks
+
+Cookbooks will now be given an official namespace, `Chef::LoadedCookbooks::NAME`, to place things in.  Resources in `resources/x.rb` (and providers in ) will be created in that namespace to avoid these conflicts.  Constant access will remain the same (since superclasses are in the constant lookup chain).
+
+Accessing such resources directly via `Chef::Resource` is deprecated.  For backwards compatibility, these resources will also be aliased into `Chef::Resource` in Chef 12, with a deprecation warning for people who subclass or initialize from that location.
+
+#### `require 'chef/resource/file'`
+
+As the resources are moving, *requires* are also moving.  In Chef 12, we will emit a deprecation warning when you require 'chef/resource/x' instead of 'chef/resources/x', but it will still work.
+
+#### Redefining `dsl_name`
+
+Redefining `dsl_name` is deprecated and will not work in 13.  In Chef 12, the base `Chef::Resource` will include a `method_defined` hook that lets us detect this, cause the DSL to change, and emit a warning.
+
+#### Creating anonymous resource classes
+
+This procedure: `Chef::Resource::MyResource = Class.new(Chef::Resource) { ... }` could be used to create a resource currently and `my_resource` would work fine in recipes.  In Chef 13, that will no longer work, because we don't do anything magical for `Chef::Resource`.  In Chef 12, we will still check for `Chef::Resource::MyResource` when `method_missing(:my_resource)` is called, and if it is found, will return it and emit a deprecation warning.
```

Comment:
I tend to like this idea, but I think that Chef::Resource::File needs to remain even in Chef 13.  I can see LWRPs being anonymous classes, but think there needs to be an API to explicitly set the class so that normal ruby inheritance can override it.